lib/docker/image_manager.js

```diff
@@ -0,0 +1,95 @@
+
+/**
+ * Image Manager is responsible for ensuring that an image is downloaded
+ * and available for a task.  A secondary goal of the image manager is to ensure
+ * that only one download of an image happens at a time.  Parallel downloads/loading
+ * of docker images has been problematic.
+ */
+export default class ImageManager {
+  /*
+   * @param {Object} runtime - Runtime object that's typically created by the worker.
+   *                           Requires a logging and docker instance.
+   */
+  constructor(runtime) {
+    assert(runtime.docker, 'Docker instance must be provided');
+    this.runtime = runtime;
+    this.docker = runtime.docker;
+    this.log = runtime.log;
+    this._lastImageEnsured = null;
+  }
+
+  /*
+   * Ensures that an image is available for a task.  This image could be an indexed
+   * image (type: indexed-image) or a docker image.
+   *
+   * Example Docker Image (backwards compatability):
+   * imageDetails = 'ubuntu:14.04'
+   *
+   * Example Docker Image:
+   * imageDetails = {
+   *   type: 'docker-image',
+   *   name: 'ubuntu:14.04'
+   * }
+   *
+   * Example Indexed Image:
+   * imageDetails = {
+   *   type: 'indexed-image',
+   *   namespace: 'public.images.ubuntu.14_04',
+   *   path: 'public/image.tar'
+   * }
+   *
+   *
+   * @param {Object|String} imageDetails - Object or string represent
+   *
+   * @returns {Object} promise - Returns a promise that either is immediately resolved
+   *                             or will be once image is downloaded.  Ensures
+   *                             pulls/downloads are done serially.
+   */
+  async ensureImage(imageDetails, stream, scopes = []) {
+    if (typeof imageDetails === 'string') {
+      imageDetails = {
+        name: imageDetails,
+        type: 'docker-image'
+      };
+    }
+
+    return this._lastImageEnsured = Promise.resolve(this._lastImageEnsured)
+      .catch(() => {}).then(async () => {
+        this.log('ensure image', {
+         image: imageDetails
+        });
+
+        let imageHandler = this.getImageHandler(imageDetails, stream, scopes);
+        let exists = await imageHandler.imageExists();
+
+        if (!exists) {
+          await imageHandler.download();
+        }
+
+        return imageHandler.imageId;
+      });
+  }
+
+  /*
+   * Creates the appropriate image handler for the 'type' requested.
+   *
+   * @param {Object} image  - Image details including 'type'
+   * @param {Object} stream - Stream object used for piping messages to the task log
+   * @param {Array}  scopes - Array of task scopes
+   */
+  getImageHandler(image, stream, scopes) {
+    let handler;
+    if (image.type === 'docker-image') {
+      handler = new DockerImage(this.runtime, image, stream, scopes);
+    } else if (image.type === 'indexed-image') {
+      handler = new IndexedImage(this.runtime, image, stream, scopes);
+    } else {
+      throw new Error(`Unrecognized image type. Image Type: ${image.type}`);
+    }
```

Comment:
So good way to structure this code is to have a base-class (even if it just documents methods), dictionary of strategies in `image_manager.js`.. Oh, an instead of `imageHandler.download` I would just make a factory.. Example:
```js
class ImageHandler extends events.EventEmitter {
  constructor() {
    this._refCount = 0;
  }
  /** gets the imageId that it's inserted with in docker */
  get _dockerImageId { throw new Error("Must be implement in subclass"); }
  /* determine if scopes grants access to the image */
  _isAuthorized(scopes) { throw new Error("Must be implement in subclass"); }
  garbageCollect() {
    if (this._refCount === 0) {
      this._remove();
      this.emit('deleted', this);
    }
  }
  /** delete resources held by this image */
  _remove() {throw new Error("Must be implement in subclass");}
  /**Use this image in `func(dockerImageId)` which may be an async function,
   * this way you never forget to decrement refCount */
  async use(func) {
    this._refCount += 1;
    try {
      await func(this._dockerImageId)
    } finally {
      this._refCount -= 1;
    }
  }
};
```

```js

const IMAGE_LOADERS = {
  'docker-image': require('./docker-image'),  // these files then export an async function that load the image 
  'indexed-image': require('./tar-image').indexedLoader, // and after loading creates a subclass of ImageHandler
  'artifact-image': require('./tar-image').artifactLoader, // that they then return...
};

class ImageManager {
  ...
  ensureImage(imageInfo, stream, scopes) {
    var id = this.hash(imageInfo);
    if (this._cache[id]) {
      return this._cache[id];
    }
    // So cache[id] is a promise, hence, two subsequent calls will never download the image twice
    // or construct the ImageHandler object twice.
    this._cache[id] = this._lastImageEnsured = Promise.resolve(this._lastImageEnsured).then(() => {
       return IMAGE_LOADERS[imageInfo.type](imageInfo, stream, scopes);
    });
    this._cache[id].then(handler => {
      handler.once('deleted', h => {
        delete this._cache[id]; // remove from cache when it's is deleted
      });
    });
    return this._cache[id];
  }
  hash(imageInfo) { /* compute hash from imageInfo */ }
};
```

```js
class TarImageHandler extends ImageHandler {
  ....
};

exports.indexedLoader = async (...) => {
  // look up taskId of indexed artifact
  exports.artifactLoader(taskId, .....); // So reuse the other logic
};

exports.indexedLoader = async (...) => {
   // download, do the tar stream things... 
  return new TarImageHandler(...);
};
```
This is just illustrative.. The notable thing is that `ImageHandler` subclasses can't have errors. They don't do anything async... They just represent a resources and guard access and garbage collection.
`ImageManager` ensures two images aren't loaded at the same time, and that there is only ever one `ImageHandler` for every image.

I think this is a better OOP design... I generally find that whenever you are about to make an async contructor, and find that an async initializer like `download()` is necessary, it's often better to just do a factory method, so all methods on the object wrapping the resource are safe.

Also I don't believe you handle scopes for cached images correctly at the moment. Ie. once image is loaded, you don't really check scopes for the next task trying to use the image. My guess is that `docker-worker` will need to know what scopes it used to download the image artifact...  I think that's acceptable as it's always the same pattern.